man/man3/MobilePhone.3
.TH "MobilePhone" 3 "Thu May 20 2021" "Simulator" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MobilePhone
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <MobilePhone\&.h>\fP
.PP
Inherits \fBHoldableAgent\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMobilePhone\fP (const \fBMap\fP *m, const unsigned long id, Point *initPosition, \fBAgent\fP *holder, const \fBClock\fP *clock, double threshold, \fBHoldableAgent::CONNECTION_TYPE\fP connType)"
.br
.ti -1c
.RI "virtual \fB~MobilePhone\fP ()"
.br
.ti -1c
.RI "const string \fBtoString\fP (bool detailed=false) const override"
.br
.ti -1c
.RI "Point * \fBmove\fP () override"
.br
.ti -1c
.RI "bool \fBtryConnect\fP () override"
.br
.ti -1c
.RI "const \fBMobileOperator\fP * \fBgetMobileOperator\fP () const"
.br
.ti -1c
.RI "void \fBsetMobileOperator\fP (\fBMobileOperator\fP *mno)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static const string \fBgetHeader\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "double \fBm_threshold\fP"
.br
.ti -1c
.RI "\fBAntenna\fP * \fBm_connectedTo\fP"
.br
.ti -1c
.RI "\fBHoldableAgent::CONNECTION_TYPE\fP \fBm_connType\fP"
.br
.ti -1c
.RI "\fBMobileOperator\fP * \fBm_mno\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
This class represents a mobile phone\&. A mobile phone is owned by a \fBPerson\fP object and it moves on the map together with its owner\&. While moving, at every time step, when a new location is set, it interacts with surrounding antennas, trying to connect to the antenna that provides the best signal in terms of signal strength/dominance, according to the handover mechanism read from the simulation configuration file\&. The connection event is triggered by the call of \fBsetLocation()\fP\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MobilePhone::MobilePhone (const \fBMap\fP * m, const unsigned long id, Point * initPosition, \fBAgent\fP * holder, const \fBClock\fP * clock, double threshold, \fBHoldableAgent::CONNECTION_TYPE\fP connType)\fC [explicit]\fP"
Builds a new \fBMobilePhone\fP object with the parameters provided by the user\&. 
.PP
\fBParameters\fP
.RS 4
\fIm\fP a pointer to the \fBMap\fP object where the simulation takes place\&. 
.br
\fIid\fP the id of the \fBMobilePhone\fP object\&. 
.br
\fIinitPosition\fP the initial location of the mobile phone on the map\&. 
.br
\fIholder\fP a pointer to the \fBAgent\fP object that owns this mobile phone\&. 
.br
\fIclock\fP a pointer to the \fBClock\fP object used in this simulation\&. 
.br
\fIthreshold\fP the minimum signal dominance or signal strength of the field below which the mobile phone cannot connect to an antenna\&. 
.br
\fIconnType\fP the type of the handover mechanism: based on the signal strength or signal dominance\&. It could take one of the following values: HoldableAgent::CONNECTION_TYPE::USING_SIGNAL_DOMINANCE or HoldableAgent::CONNECTION_TYPE::USING_SIGNAL_STRENGTH\&. 
.RE
.PP

.SS "virtual MobilePhone::~MobilePhone ()\fC [virtual]\fP"
The default destructor of the class\&. 
.SH "Member Function Documentation"
.PP 
.SS "static const string MobilePhone::getHeader ()\fC [static]\fP"
Returns a string with the header line of the description of the mobile phone returned by the \fBtoString()\fP method\&. It is a static method, the header line being the same for all mobile phones regardless the owner\&. 
.PP
\fBReturns\fP
.RS 4
a string with the header line of the description of the mobile phone returned by the \fBtoString()\fP method\&. 
.RE
.PP

.SS "const \fBMobileOperator\fP* MobilePhone::getMobileOperator () const"
Returns the \fBMobileOperator\fP object that owns this mobile phone\&. Each mobile phone should belong to a Mobile Operator\&. 
.PP
\fBReturns\fP
.RS 4
the \fBMobileOperator\fP object that owns this mobile phone\&. Each mobile phone should belong to a Mobile Operator\&. 
.RE
.PP

.SS "Point* MobilePhone::move ()\fC [inline]\fP, \fC [override]\fP, \fC [virtual]\fP"
Makes a step on the map according to the movement strategy set to the \fBPerson\fP object which owns this mobile phone\&. The displacement strategy is set at the \fBPerson\fP object creation and currently the following strategies are supported: \fBRandomWalkDisplacement\fP, \fBRandomWalkDriftDisplacement\fP, \fBLevyFlightDisplacement\fP and \fBHomeWorkDisplacement\fP\&. 
.PP
\fBReturns\fP
.RS 4
the final location after the displacement\&. The mobile phone moves together with the \fBPerson\fP which owns it\&. 
.RE
.PP

.PP
Implements \fBMovableAgent\fP\&.
.SS "void MobilePhone::setMobileOperator (\fBMobileOperator\fP * mno)"
Sets the \fBMobileOperator\fP object which owns this mobile phone\&. 
.PP
\fBParameters\fP
.RS 4
\fImno\fP a pointer to the \fBMobileOperator\fP object which owns this mobile phone\&. 
.RE
.PP

.SS "const string MobilePhone::toString (bool detailed = \fCfalse\fP) const\fC [override]\fP, \fC [virtual]\fP"
Builds a string with the most relevant information of the class\&. It is useful to output the description of the mobile phone to the console or to a file\&. The parameter
.PP
\fBParameters\fP
.RS 4
\fIis\fP currently ignored\&. 
.br
\fIdetailed\fP is currently ignored\&. 
.RE
.PP
\fBReturns\fP
.RS 4
a string object containing the id of the mobile phone, the coordinates of the location, the current speed of movement, the id of the \fBPerson\fP object that owns this mobile phone and the id of the mobile network operator where this mobile phone has a subscription\&. 
.RE
.PP

.PP
Reimplemented from \fBHoldableAgent\fP\&.
.SS "bool MobilePhone::tryConnect ()\fC [override]\fP, \fC [virtual]\fP"
This method is called after the mobile phone moves (together with its owner) to a new location\&. It tries to connect the mobile phone to an antenna\&. The connection method is determined by inspecting the \fIm_connType\fP member: using the signal dominance or using the signal strength\&. The value of the \fIm_connType\fP member is set by the constructor of the class\&. If the connection is successfully, a pointer to the \fBAntenna\fP object where this mobile phone was connected is stored internally\&. First, if the mobile phone is already connected to an antenna, this method checks if the mobile phone is still providing a signal strong enough for a connection and if not it detaches from that antenna\&. Then it selects the antenna providing the best signal (in terms of signal strength or dominance) and tries to connect to it\&. If the connection is successful it saves a pointer to that \fBAntenna\fP object\&. If this antenna reject the connection( for example it reached its maximum number of simultaneous connections) then a list of all other antennas is build, ordered by the signal strength/dominance\&. The mobile phone iterates through this list and tries to connect to an antenna\&. If one antenna accepts the incoming connection the method returns true, otherwise it returns false\&. When a connection is established, the pointer to the \fBAntenna\fP object is updated accordingly\&. 
.PP
\fBReturns\fP
.RS 4
true if the connection succeeds, false otherwise\&. 
.RE
.PP

.PP
Implements \fBHoldableAgent\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBAntenna\fP* MobilePhone::m_connectedTo\fC [private]\fP"

.SS "\fBHoldableAgent::CONNECTION_TYPE\fP MobilePhone::m_connType\fC [private]\fP"

.SS "\fBMobileOperator\fP* MobilePhone::m_mno\fC [private]\fP"

.SS "double MobilePhone::m_threshold\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for Simulator from the source code\&.


src/agent/MobilePhone.cpp
/*
 * Copyright (C) 2019  Bogdan Oancea

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version and under the EUPL free software license version 1.0 or later.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/> and
 * <https://ec.europa.eu/info/european-union-public-licence_en>
 *
 * A data simulator for mobile phone network events
 *
 * MobilePhone.cpp
 *
 *  Created on: Apr 5, 2019
 *      Author: Bogdan Oancea
 *      Email : bogdan.oancea@gmail.com
 */

#include <agent/Antenna.h>
#include <agent/MobileOperator.h>
#include <agent/MobilePhone.h>
#include <EMField.h>
#include <iomanip>
#include <sstream>
#include <utility>
#include <vector>

using namespace std;

MobilePhone::MobilePhone(const Map* m, const unsigned long id, Point* initPosition, Agent* holder, const Clock* clock, double threshold, HoldableAgent::CONNECTION_TYPE connType) :
		HoldableAgent(m, id, initPosition, nullptr, clock), m_threshold { threshold }, m_connType { connType } {
	m_connectedTo = nullptr;
	m_mno = nullptr;
}

MobilePhone::~MobilePhone() {
}

const string MobilePhone::toString(bool detailed) const {
	ostringstream result;

	result << HoldableAgent::toString();
	if (m_mno)
		result << left << setw(15) << m_mno->getId();
	else
		result << left << setw(15) << "No MNO";
	return (result.str());
}

bool MobilePhone::tryConnect() {
	bool connected = false;
	//select the most powerful antenna
	Point *p = this->getLocation();
	if (m_connectedTo != nullptr) {
		//check if this antenna is still in range, otherwise detach from it
		bool inRange = EMField::instance()->isAntennaInRange(p, m_connectedTo, m_threshold, m_connType);
		if (!inRange) {
			m_connectedTo->dettachDevice(this);
			m_connectedTo = nullptr;
			connected = false;
		}
	}

	pair<Antenna*, double> antenna;
	if (m_connType == HoldableAgent::CONNECTION_TYPE::USING_POWER)
		antenna = EMField::instance()->computeMaxPower(p, getMobileOperator()->getId());
	else if (m_connType == HoldableAgent::CONNECTION_TYPE::USING_SIGNAL_DOMINANCE) {
		antenna = EMField::instance()->computeMaxQuality(p, getMobileOperator()->getId());
	} else if (m_connType == HoldableAgent::CONNECTION_TYPE::USING_SIGNAL_STRENGTH) {
		antenna = EMField::instance()->computeMaxStrength(p, getMobileOperator()->getId());
	}

	if (antenna.first != nullptr && antenna.second > m_threshold) {
		connected = antenna.first->tryRegisterDevice(this);
	}
	if (connected) {
		if (m_connectedTo != nullptr && antenna.first->getId() != m_connectedTo->getId()) {
			m_connectedTo->dettachDevice(this);
		}
		m_connectedTo = antenna.first;
	} else {
		//try to connect to another antenna in range
		//antennas need to belong to the same MNO
		vector<pair<Antenna*, double>> antennas = EMField::instance()->getInRangeAntennas(p, m_threshold, m_connType, getMobileOperator()->getId());
		unsigned long size = antennas.size();
		for (unsigned long i = 0; i < size; i++) {
			connected = antennas[i].first->tryRegisterDevice(this);
			if (connected) {
				if (m_connectedTo != nullptr && antennas[i].first->getId() != m_connectedTo->getId()) {
					m_connectedTo->dettachDevice(this);
				}
				m_connectedTo = antennas[i].first;
				break;
			}
		}
	}
	return (connected);
}


const MobileOperator* MobilePhone::getMobileOperator() const {
	return (m_mno);
}

void MobilePhone::setMobileOperator(MobileOperator* mno) {
	m_mno = mno;
}

const string MobilePhone::getHeader()  {
	ostringstream result;
	result << left << setw(15) << "Phone ID" << setw(15) << " X " << setw(15) << " Y " << setw(15) << " Speed " << setw(15) << " Owner id " << setw(15) << "MNO Id" << endl;
	return result.str();
}


